api/search/queryprocessor.py

```diff
@@ -158,19 +158,22 @@ def get_results(self, parent_name, parent_results):
 
 class PreparedSearch(object):
 
-    containers = ['groups', 'projects', 'sessions', 'collections', 'acquisitions']
-
     def __init__(self, target_paths, queries, all_data=False, user=None):
+        self.containers = ['groups', 'projects', 'sessions', 'collections', 'acquisitions']
         self.queries = queries
         self.target_lists = {}
         for path in target_paths:
             targets = self._get_targets(path)
             self._merge_into(targets, self.target_lists)
         self.search_containers = {}
-        for cont_name in self.containers:
+        last_relevant_container = None
+        for i, cont_name in enumerate(self.containers[::-1]):
             query = self.queries.get(cont_name)
             targets = self.target_lists.get(cont_name, [])
-            self.search_containers[cont_name] = SearchContainer(cont_name, query, targets, all_data, user)
+            if query or targets or last_relevant_container is not None:
+                last_relevant_container = last_relevant_container if last_relevant_container is not None else i
+                self.search_containers[cont_name] = SearchContainer(cont_name, query, targets, all_data, user)
+        self.containers = self.containers[:None if last_relevant_container == 0 else - last_relevant_container]
```

Comment:
I think this might be cleaner if written as 
```
if last_relevant_container is None:
    last_relevant_container = i
```

It's easier to read if you're only setting that variable when necessary. It took me a while to figure out what was going on here as written. I would also add a comment or two about this loop and the line following for clarity.